### PR TITLE
Ssl for postgres

### PR DESCRIPTION
**Pull request by ramn**
Fixes https://github.com/Keats/dbmigrate/issues/6

Tested and works for me.

Parses sslmode as specified here: http://www.postgresql.org/docs/current/static/libpq-ssl.html#LIBPQ-SSL-SSLMODE-STATEMENTS
But no support for verification.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,11 +12,13 @@ keywords = ["database", "postgres", "migration", "sql", "mysql"]
 [dependencies]
 regex = "0.1"
 url = "0.5"
-postgres = "0.11"
+postgres = { version = "0.11.3", features = ["openssl"] }
```
Comment (Keats) [inline Cargo.toml]: keep 0.11 as version so we get the minor updates (postgres is currently 0.11.4 for example)

File: Cargo.toml [hunk 2]
```diff
@@ -12,11 +12,13 @@ keywords = ["database", "postgres", "migration", "sql", "mysql"]
...
 mysql = "1.2"
 clippy = {version = "~0.0.48", optional = true}
 time = "0.1"
 term = "0.2"
+openssl = "0.7.6"
+openssl-sys = "0.7.6"
```
Comment (Keats) [inline Cargo.toml]: same as above don't put the bug fix version (so it will be "0.7") for both

File: src/drivers/postgres.rs [hunk 3]
```diff
@@ -56,3 +58,30 @@ impl Driver for Postgres {
         Ok(())
     }
 }
+
+fn mk_connection(url: &str) -> Result<Connection, MigrateError> {
```
Comment (Keats) [inline src/drivers/postgres.rs]: use a `MigrateResult` here.

File: src/drivers/postgres.rs [hunk 4]
```diff
@@ -56,3 +58,30 @@ impl Driver for Postgres {
...
+    let ctx = try!(SslContext::new(SslMethod::Sslv23));
```
Comment (Keats) [inline src/drivers/postgres.rs]: this is created even when we don't have ssl mode, which is a bit unfortunate

File: src/drivers/postgres.rs [hunk 5]
```diff
@@ -56,3 +58,30 @@ impl Driver for Postgres {
...
+            pairs.into_iter().find(|&(ref k, _)| k == SSLMODE))
+        .map(|(_, v)| match v.as_str() {
+            "allow" | "prefer" => SslMode::Prefer(&ctx),
+            "require" => SslMode::Require(&ctx),
+            // No support for certificate verification yet.
+            "verify-ca" | "verify-full" => unimplemented!(),
+            _ => SslMode::None })
+        .unwrap_or(SslMode::None);
+
+    Connection::connect(without_sslmode(&url).as_ref(), sslmode)
```
Comment (Keats) [inline src/drivers/postgres.rs]: can you match on sslmode here instead?

File: src/drivers/postgres.rs [hunk 6]
```diff
@@ -56,3 +58,30 @@ impl Driver for Postgres {
...
+        .map_err(From::from)
+}
+
+fn without_sslmode(url: &Url) -> String {
```
Comment (Keats) [inline src/drivers/postgres.rs]: does rust-postgres fail if you provide the sslmode in the url string ?

Comment (ramn) [inline src/drivers/postgres.rs]: Yes

File: src/drivers/postgres.rs (hunk 5 above)
Comment (ramn) [inline src/drivers/postgres.rs]: Instead of what?

File: src/drivers/postgres.rs (hunk 4 above)
Comment (ramn) [inline src/drivers/postgres.rs]: Aesthetically speaking a bit unfortunate, however it happens just once and is not a performance problem. I was thinking that it isn't worth complicating the code.

File: src/drivers/postgres.rs (hunk 6 above)
Comment (Keats) [inline src/drivers/postgres.rs]: hm that's a bit annoying, i'll ask on the rust-postgres repo if that's intended

File: src/drivers/postgres.rs (hunk 5 above)
Comment (Keats) [inline src/drivers/postgres.rs]: nvm that was a bad idea

**Comment by Keats** [comment]
rust-posgres is not going to support automatic ssl from the url so i'll merge that and release later today.

Thanks!

**Comment by Keats** [comment]
Published 0.2.3 that contains that and your previous PR, let me know if everything works ok!

**Comment by ramn** [comment]
Thanks yes works fine!

**Pull request merged**